racket/collects/racket/hash.rkt

```diff
@@ -63,6 +63,23 @@
                     (combine/key k v (hash-ref hm k))))
         res)))
 
+(define (hash-filter ht pred)
+  (cond [(immutable? ht)
+         (for ([(k v) (in-hash ht)])
+           (when (not (pred k v))
+             (set! ht (hash-remove ht k))))]
+        [else
+         (for ([(k v) (in-hash ht)])
```

Comment:
As @AlexKnauth said, let's make a new hash table via `hash-copy-clear`, and then add stuff to it, so that we don't mutate the original hash table.